### PR TITLE
fix(components): Fix state warning errors

### LINKED ISSUE #1243: Console Errors when add certain components
## Detailed description
[
screen shot 2018-08-28 at 11 02 21
](url)

When My application starts I see console error messages for some components for
 - AccordionItem
 - PaginationV2
 - Tooltip
 - NumberInput
 - ProgressIndicator
E.g
warning.js:33 Warning: ProgressIndicator: Did not properly initialize state during construction. Expected state to be an object, but it was undefined.

> Describe in detail the issue you're having. The Component seems to behave normally but i am getting stack traces on my browser console since upgrading to ^6.26.0 from 6.14.3

> Is this a feature request (new component, new icon), a bug, or a general issue? general - I dont see it in all components we are using e.g TooltipIcon

> Is this issue related to a specific component? Only seen it in some components

> What did you expect to happen? What happened instead? What would you like to see changed? Console was clear of errors

> What browser are you working in? Chrome mainly

> What version of the Carbon Design System are you using? ^6.26.0

> What offering/product do you work on? Any pressing ship or release dates we should be aware of?
Transformation Advisor
## Steps to reproduce the issue

1. Step one
2. Step two
3. Step three
4. etc.

## Additional information

* Screenshots or code
* Notes

### PR DESCRIPTION
**Pull request by zerostatic**
Closes #1243

A few components are giving the following error in the console:
```
Did not properly initialize state during construction. 
Expected state to be an object, but it was undefined.
```

#### Changelog

The following components were fixed:
AccordionItem
ComposedModal
ContentSwitcher
FileUploader
NumberInput
OverflowMenu
PaginationV2
ProgressIndicator
RadioButtonGroup
SearchLayoutButton
TimePicker.js
Tooltip.js

#### Notes

A few components were checking is the state variable was defined. I had to change this now to check if state is an empty object. This is the instances were you see this: 
```
let emptyState =
      Object.keys(state).length === 0 && state.constructor === Object;
```

giphy

### CODE REVIEW

**Comment by liviudobrea** [comment]
There's also this https://github.com/facebook/react/issues/12523
Personally, I would rather have it explicitly defined, so I wouldn't rely on gDFSP setting up initial state for me.
I don't expect React to add a fallback for when state is null on first instance and gDFSP, rather they'll probably suppress the error at some point with a warning. But that doesn't fix the initial issue.
That being said, I'm with @zerostatic on this one.

**Comment by asudoh** [comment]
Weighing in - My earlier comment is not about coding preference, but about not breaking existing code. This PR need more work in this regard. Thanks!

**Comment by asudoh** [comment]
Rewind - Looking at `gDSFP()` logic again I see a breakage won’t happen. My apologies for the false alarm. @zerostatic Do you want to update `gDSFP()` logic to remove the obsolete code handling undefined `state`? If not that’s fine and I can do that afterwards. Thanks!

**Comment by asudoh** [comment]
Rewind - Looking at `gDSFP()` logic again I see a breakage won’t happen. My apologies for the false alarm. @zerostatic Do you want to update `gDSFP()` logic to remove the obsolete code handling undefined state? If not that’s fine and I can do that afterwards. Thanks!

**Comment by zerostatic** [comment]
@asudoh I'm not sure I understand, 
Now that I've added an empty state object: `state = {}`
I replaced the code that checks for undefined state with code that checks if state is an empty object, like:

I replaced this:
```if (!state) {...```

with this:
``` if (Object.keys(state).length === 0 && state.constructor === Object) {....```

Is there a cleaner way to handle this?

**Comment by asudoh** [comment]
@zerostatic Thank you for asking and my apologies for the confusion. Let’s take your change to `AccordionItem.js` as an example - Given you removed “undefined `state`” condition, we can remove the logic handling that condition in `gDSFP()`, like:

```diff
--- AccordionItem.js.orig	2018-09-03 06:36:01.000000000 +0900
+++ AccordionItem.js	2018-09-03 06:36:13.000000000 +0900
@@ -51,8 +51,8 @@
   };
 
   static getDerivedStateFromProps({ open }, state) {
-    const { prevOpen } = state || {};
-    return state && prevOpen === open
+    const { prevOpen } = state;
+    return prevOpen === open
       ? null
       : {
           open,
```

Don’t hesitate to ping me if you have any further questions - Thanks!

**Comment by zerostatic** [comment]
@asudoh Thanks for the explanation! I've made the updates. Let me know if I can do anything to improve it.

**Comment by asudoh** [comment]
Thanks @zerostatic for your updates! Would you want to make similar changes to `gDSFP()` in files where you added empty initial states? If you don’t feel like it it’ll be no problem - Let us know. Thanks!

**Comment by zerostatic** [comment]
@asudoh No problem...the other gDSFP's have been updated.

File: src/components/NumberInput/NumberInput.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ import Icon from '../Icon';
 import classNames from 'classnames';
 
 export default class NumberInput extends Component {
+  state = { value: '' };
```
[the remark this answers is not in the document]
Comment (zerostatic) [reply src/components/NumberInput/NumberInput.js]: Sure thing...this has been updated. Thanks!

**Pull request merged**